consensus/core/src/core.rs

```diff
@@ -262,12 +273,34 @@ impl Core {
             let _ = self.signals.new_block_ready(verified_block.reference());
             // TODO: propagate shutdown or ensure this will never return error?
 
+            self.try_commit();
```

Comment:
As discussed earlier, this should be called from `add_blocks()`. Or leave a TODO.

Reply:
`try_new_block()` is called by `add_blocks()` and `force_new_block()`, so it seemed to make sense to just have `try_commit()` in one location?

Reply:
There are a number of early return conditions so `try_commit()` may not get called even if commit is possible. In general commit is independent of propose.

Reply:
I agree on the early returns of the `try_new_block` method. Just note though that given how the universal_committer/base_committer are structured, we'll attempt to apply both the `direct` and `indirect` commit rules for every single block we receive for the decision round - and we know that we can't apply the direct rule if we don't have at least `2f+1` blocks at decision round. So may I suggest moving it to the `add_block` but call it once a new quorum gets formed? Alternatively we just move it to the add blocks method but let's keep that under our radar in case it's proven to be a bit expensive.

Reply:
Yes this optimization should be implemented as discussed in a previous PR. But the logic can be: after adding a block, if `R` is the last round with a quorum, run commit rule if there is any undecided leader `<= R-2`. Further filtering based on block round is possible too. Eventually per round quorum status can be cached and even shared with the propose logic. I guess if this is too much change for this PR, it is ok to leave a TODO. But we really should revisit this.

Reply:
Sorry I did see your comment in the other PR, just failed to add it correctly. I left a TODO and will follow up with the optimization in another PR after our initial private-testnet run.